examples/caa/caa.ipynb

```diff
@@ -0,0 +1,766 @@
+      "Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.\n",
+      "HTTP request sent, awaiting response... 200 OK\n",
+      "Length: 11126 (11K) [application/octet-stream]\n",
+      "Saving to: ‘vec_layer_15_Llama-2-13b-chat-hf.pt.1’\n",
+      "\n",
+      "vec_layer_15_Llama- 100%[===================>]  10.87K  --.-KB/s    in 0s      \n",
+      "\n",
+      "2024-01-28 11:45:53 (99.9 MB/s) - ‘vec_layer_15_Llama-2-13b-chat-hf.pt.1’ saved [11126/11126]\n",
+      "\n"
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...\n",
+      "To disable this warning, you can either:\n",
+      "\t- Avoid using `tokenizers` before the fork if possible\n",
+      "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Download the CAA sycophancy vectors for layer 15\n",
+    "!wget https://raw.githubusercontent.com/nrimsky/CAA/main/vectors/vec_layer_15_Llama-2-7b-chat-hf.pt\n",
+    "!wget https://raw.githubusercontent.com/nrimsky/CAA/main/vectors/vec_layer_15_Llama-2-13b-chat-hf.pt"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 16,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Cosine similarity: 0.995\n"
+     ]
+    }
+   ],
+   "source": [
+    "from torch.nn.functional import cosine_similarity\n",
+    "\n",
+    "original_steering_vector = torch.load(f'vec_layer_15_Llama-2-{model_size}-chat-hf.pt')\n",
+    "our_steering_vector = steering_vector.layer_activations[15]\n",
+    "print(f\"Cosine similarity: {cosine_similarity(original_steering_vector, our_steering_vector, dim=0):.3f}\")\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Steer with Steering Vectors"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### An Introduction to PyTorch Hooks\n",
+    "To apply steering vectors to the model, we must have some way of modifying the model's forward pass. The approach taken in the official codebases for Representation Engineering and Contrastive Activation Addition both do this by wrapping the model's blocks. This approach is conceptually simpler but has notable limitations, e.g. being unable to work for other models\n",
+    " \n",
+    "The `steering_vectors` library uses Pytorch hooks instead of model wrappers to modify the underlying model's forward pass. You can read more about hooks at the [official PyTorch documentation](https://pytorch.org/docs/stable/generated/torch.nn.modules.module.register_module_forward_hook.html). \n",
```

Comment:
This is personal opinion, but I feel like users almost never care about how a library works as long as it works. Most users won't look at what CAA is doing in their repo either, or really be interested in the differences between these things, and if they are the source code is in Github anyway and isn't that complex. I think this stuff is interesting to us since we've thought a lot about how this stuff works internally and why, but users generally just want a magic black box that is as easy to use as possible, without needing to think about what's under the hood.

We could add a section about Pytorch hooks and how this works and why to the `about` page of the docs potentially, since that has info about how to contribute and might be more relevant for people wanting to do development on the library.

Reply:
Alright fair enough, I'll get rid of this section for now.